src/level_four/tap_cards/controller/card_action_controllers.c

```diff
@@ -0,0 +1,168 @@
+ *          This file contains the functions to retrieve wallet from cards.
+ * @copyright Copyright (c) 2022 HODL TECH PTE LTD
+ * <br/> You may obtain a copy of license at <a href="https://mitcc.org/" target=_blank>https://mitcc.org/</a>
+ * 
+ ******************************************************************************
+ * @attention
+ *
+ * (c) Copyright 2022 by HODL TECH PTE LTD
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining
+ * a copy of this software and associated documentation files (the
+ * "Software"), to deal in the Software without restriction, including
+ * without limitation the rights to use, copy, modify, merge, publish,
+ * distribute, sublicense, and/or sell copies of the Software, and to
+ * permit persons to whom the Software is furnished to do so, subject
+ * to the following conditions:
+ * 
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *  
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR
+ * ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF
+ * CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+ * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *  
+ *  
+ * "Commons Clause" License Condition v1.0
+ *  
+ * The Software is provided to you by the Licensor under the License,
+ * as defined below, subject to the following condition.
+ *  
+ * Without limiting other conditions in the License, the grant of
+ * rights under the License will not include, and the License does not
+ * grant to you, the right to Sell the Software.
+ *  
+ * For purposes of the foregoing, "Sell" means practicing any or all
+ * of the rights granted to you under the License to provide to third
+ * parties, for a fee or other consideration (including without
+ * limitation fees for hosting or consulting/ support services related
+ * to the Software), a product or service whose value derives, entirely
+ * or substantially, from the functionality of the Software. Any license
+ * notice or attribution required by the License must also include
+ * this Commons Clause License Condition notice.
+ *  
+ * Software: All X1Wallet associated files.
+ * License: MIT
+ * Licensor: HODL TECH PTE LTD
+ *
+ ******************************************************************************
+ */
+#include "controller_tap_cards.h"
+#include "flash_api.h"
+#include "nfc.h"
+#include "shamir_wrapper.h"
+#include "tasks.h"
+#include "ui_instruction.h"
+#include "ui_message.h"
+#include "wallet_utilities.h"
+#include "wallet.h"
+
+static void handle_retrieve_wallet_success(uint8_t xcor);
+
+void readback_share_from_card(uint8_t xcor)
+{
+    tap_card_data.retries = 5;
```

Comment:
Should check card presence here to avoid inconsistent message

Reply:
i do not understand this. what msg do you see which should be prevented?

Reply:
If you remove the card in between "write to card" and "read from card", the screen "card detected, processing..." stays on the screen which is incosistent

Reply:
added